pkg/proxy/proxy.go

```diff
@@ -114,39 +133,61 @@ func createProxy(
 	envoyIntegration *envoyProxyIntegration,
 	dnsIntegration *dnsProxyIntegration,
 ) *Proxy {
-	return &Proxy{
+	p := &Proxy{
 		rangeMin:         minPort,
 		rangeMax:         maxPort,
 		redirects:        make(map[string]*Redirect),
 		datapathUpdater:  datapathUpdater,
 		allocatedPorts:   make(map[uint16]bool),
 		proxyPorts:       defaultProxyPortMap(),
+		proxyPortsPath:   filepath.Join(option.Config.StateDir, proxyPortsFile),
 		envoyIntegration: envoyIntegration,
 		dnsIntegration:   dnsIntegration,
 	}
+
+	triggerDone := make(chan struct{})
+	p.triggerDone = triggerDone
+	p.proxyPortsTrigger, _ = trigger.NewTrigger(trigger.Parameters{
+		MinInterval:  10 * time.Second,
+		TriggerFunc:  p.storeProxyPorts,
+		ShutdownFunc: func() { close(triggerDone) },
+	})
+
+	return p
 }
 
-func defaultProxyPortMap() map[string]*ProxyPort {
-	return map[string]*ProxyPort{
+func (p *Proxy) close() {
```

Comment:
some questions regarding this function

- is `close` really a good name?
- IMO it's not necessary to nil-check on proxyPortsTrigger as this should always be initialized (and the stop hook not executed if otherwise)
- Is it necessary to reset the trigger to `nil`?
- Why is it necessary to lock the mutex?

Reply:
I guess it is good to shut down the trigger, maybe otherwise it would be possible for the agent to exit halfway through writing the file?

I'll rename `close` to `triggerShutdown` and remove `nil`ling and `nil` checks. Also it looks like `trigger.Shutdown` is thread safe, and `trigger.Trigger` locks internally, so locking not needed here. I likely copy/pasted the close from another Trigger use in the code base without thinking too hard...

Reply:
Left the nil check after I moved trigger management to the hive lifecycle hooks.

Reply:
Removed the `nil` check by improving unit test code to add a no-op trigger.

Reply:
thanks for the refactoring!